src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1498 @@
+        SD_BUS_VTABLE_START(0),
+
+        SD_BUS_METHOD_WITH_ARGS("ListTargets",
+                                SD_BUS_NO_ARGS,
+                                SD_BUS_RESULT("a(sso)", targets),
+                                method_list_targets,
+                                SD_BUS_VTABLE_UNPRIVILEGED),
+
+        SD_BUS_SIGNAL_WITH_ARGS("JobRemoved",
+                                SD_BUS_ARGS("u", job_id, "o", job_path, "i", exit_code),
+                                0),
+
+        SD_BUS_VTABLE_END,
+};
+
+static const BusObjectImplementation manager_object = {
+        "/org/freedesktop/sysupdate1",
+        "org.freedesktop.sysupdate1.Manager",
+        .vtables = BUS_VTABLES(manager_vtable),
+        .children = BUS_IMPLEMENTATIONS(&job_object, &target_object),
+};
+
+static int manager_add_bus_objects(Manager *m) {
+        int r;
+
+        assert(m);
+
+        r = bus_add_implementation(m->bus, &manager_object, m);
+        if (r < 0)
+                return r;
+
+        r = bus_log_control_api_register(m->bus);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_request_name_async(m->bus, NULL, "org.freedesktop.sysupdate1", 0, NULL, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to request name: %m");
+
+        r = sd_bus_attach_event(m->bus, m->event, 0);
+        if (r < 0)
+                return log_error_errno(r, "Failed to attach bust to event loop: %m");
+
+        return 0;
+}
+
+static bool manager_check_idle(void *userdata) {
+        Manager *m = ASSERT_PTR(userdata);
+        return hashmap_isempty(m->jobs);
+}
+
+static int manager_run(Manager *m) {
+        assert(m);
+
+        return bus_event_loop_with_idle(m->event,
+                                        m->bus,
+                                        "org.freedesktop.sysupdate1",
+                                        DEFAULT_EXIT_USEC,
+                                        manager_check_idle,
+                                        m);
+}
+
+static int run(int argc, char *argv[]) {
+        _cleanup_(manager_unrefp) Manager *m = NULL;
+        int r;
+
+        log_setup();
+
+        r = service_parse_argv("systemd-sysupdated.service",
+                               "System update management service.",
+                               BUS_IMPLEMENTATIONS(&manager_object,
+                                                   &log_control_object),
+                               argc, argv);
+        if (r <= 0)
+                return r;
+
+        umask(0022);
+
+        assert_se(sigprocmask_many(SIG_BLOCK, NULL, SIGCHLD, SIGTERM, SIGINT, SIGRTMIN+18, -1) >= 0);
+
+        r = manager_new(&m);
+        if (r < 0)
+                return log_error_errno(r, "Failed to allocate manager object: %m");
+
+        r = manager_add_bus_objects(m);
+        if (r < 0)
+                return r;
+
+        r = manager_enumerate_targets(m);
+        if (r < 0)
+                return r;
```

Comment:
caching these things is always problematic. I think it would be nicer to cache on-the-fly, i.e. right before we need it, and then destroy the list automatically when we are idle (from an idle event source), and reread when we need it next.

after all we allow people to drop in new images if they like any time.

Reply:
I was thinking since the service quits on idle it's not really an issue

As far as I can tell that's the same approach as taken by machined